## Part 2: Product matching


- Attached is sample product lists from Flipkart & Amazon 
    <DS - Assignment Part 2 data set.zip>

- Link: https://www.dropbox.com/sh/aypq6h3254207bs/AACzMLvo-XtK9sYAAma6FW0la?dl=0

### Problem statement:


- Using ML/DL techniques, match similar products from the Flipkart dataset with the Amazon dataset. Once
similar products are matched, display the retail price from FK and AMZ side by side. Please explore as
many techniques as possible before choosing the final technique.
You may either display the final result in single table format OR You may create a simple form where we
input the product name and the output of prices of the product from both websites are displayed.
For example:
FDT Women's Leggings from Flipkart should be matched with FDT WOMEN'S Leggings Pants from
Amazon

In [1]:
import numpy as np
import pandas as pd

from warnings import filterwarnings
filterwarnings("ignore")

In [2]:
amazon = pd.read_csv("../data/amz_com_ecommerce_sample.csv", encoding = "ISO-8859-1")
flipkart = pd.read_csv("../data/flipkart_com_ecommerce_sample.csv", encoding = "ISO-8859-1")

In [3]:
amazon.sample(5)

uniq_id            crawl_timestamp  \
15083  d9d338bfba2a4dc17aa29ba754936a6b  2015-12-12 11:46:53 +0000   
4971   5347a5510f72505aac3e503d366ea324  2015-12-01 12:40:44 +0000   
6979   c36f5b14d3d88fa15447b1b957fa887f  2015-12-01 06:13:00 +0000   
12219  05457d1aff900c23bce6b1be628f8a4c  2015-12-30 00:17:46 +0000   
14232  5dca8b0c2c1e49bd2964a01b9a6c94c8  2015-12-13 00:29:55 +0000   

                                             product_url  \
15083  http://www.flipkart.com/avarnas-women-s-camiso...   
4971   http://www.flipkart.com/khoobee-silk-self-desi...   
6979   http://www.flipkart.com/falconoverseas-classy-...   
12219  http://www.flipkart.com/bpt-women-s-printed-ca...   
14232  http://www.flipkart.com/ninja-turtles-printed-...   

                                            product_name  \
15083                           Avarnas Women's Camisole   
4971   Khoobee Silk Self Design, Embroidered Blouse M...   
6979         Falconoverseas Classy Appeal Brass Necklace   
12219                   BPT Women's Printed Casual Shirt   
14232     Ninja Turtles Printed Men's Round Neck T-Shirt   

                                   product_category_tree               pid  \
15083  ["Clothing >> Women's Clothing >> Lingerie, Sl...  CSPEYCFQHPFCEQ23   
4971   ["Clothing >> Women's Clothing >> Ethnic Wear ...  FABEAAXUASTJUUFH   
6979    ["Jewellery >> Necklaces & Chains >> Necklaces"]  NKCE9ABEZP4QYJHM   
12219  ["Clothing >> Women's Clothing >> Western Wear...  SHTEBGHFMTUYHDYH   
14232  ["Clothing >> Men's Clothing >> T-Shirts >> Ni...  TSHE4CDEGAPTFXVK   

       retail_price  discounted_price  \
15083           184               237   
4971           2448              1169   
6979            589               351   
12219          1039               471   
14232           739               351   

                                                   image  \
15083  ["http://img6a.flixcart.com/image/camisole-sli...   
4971   ["http://img6a.flixcart.com/image/fabric/h/x/m...   
6979   ["http://img5a.flixcart.com/image/necklace-cha...   
12219  ["http://img6a.flixcart.com/image/shirt/d/y/h/...   
14232  ["http://img6a.flixcart.com/image/t-shirt/d/z/...   

       is_FK_Advantage_product  \
15083                    False   
4971                     False   
6979                     False   
12219                    False   
14232                     True   

                                             description       product_rating  \
15083  Avarnas Women's Camisole\n                    ...  No rating available   
4971   Buy Khoobee Silk Self Design, Embroidered Blou...  No rating available   
6979   Falconoverseas Classy Appeal Brass Necklace - ...  No rating available   
12219  BPT Women's Printed Casual Shirt\n            ...  No rating available   
14232  Ninja Turtles Printed Men's Round Neck T-Shirt...  No rating available   

            overall_rating           brand  \
15083  No rating available             NaN   
4971   No rating available             NaN   
6979   No rating available  Falconoverseas   
12219  No rating available            Slim   
14232  No rating available             NaN   

                                  product_specifications  
15083  {"product_specification"=>[{"key"=>"Number of ...  
4971   {"product_specification"=>[{"key"=>"Number of ...  
6979   {"product_specification"=>[{"key"=>"Collection...  
12219  {"product_specification"=>[{"key"=>"Pattern", ...  
14232  {"product_specification"=>[{"key"=>"Sleeve", "...

In [4]:
flipkart.sample(5)

uniq_id            crawl_timestamp  \
4787   ba807fb7fc9408fa93fe4cf703664442  2015-12-01 12:40:44 +0000   
10573  a8a8ddbe9517ca8388e0275de334f83d  2016-04-11 09:17:31 +0000   
6155   d9a46fc187c6265750410f830a27e401  2015-12-01 06:13:00 +0000   
2661   ba320ce7ee7b90d6648b462371cffd00  2016-01-07 06:04:02 +0000   
13600  b7060956bc0ded3f0a4c2a2ea336ecf4  2016-01-01 19:37:22 +0000   

                                             product_url  \
4787   http://www.flipkart.com/dailyware-tasara-kadha...   
10573  http://www.flipkart.com/vastram-casual-sleevel...   
6155   http://www.flipkart.com/joyeria-milan-resin-ne...   
2661   http://www.flipkart.com/s4s-women-s-full-cover...   
13600  http://www.flipkart.com/bluestone-suneha-yello...   

                                            product_name  \
4787                       Dailyware Tasara Kadhai 2.5 L   
10573         VASTRAM Casual Sleeveless Solid Girl's Top   
6155                        Joyeria Milan Resin Necklace   
2661                       S4S Women's Full Coverage Bra   
13600  BlueStone The Suneha Yellow Gold Diamond, Ruby...   

                                   product_category_tree               pid  \
4787   ["Kitchen & Dining >> Cookware >> Pots & Pans ...  PTPDZYCEXGZ8SKT2   
10573  ["Clothing >> Kids' Clothing >> Girls Wear >> ...  TOPEHFYBVAUNRGX3   
6155    ["Jewellery >> Necklaces & Chains >> Necklaces"]  NKCE3JRM4S94YMHD   
2661   ["Clothing >> Women's Clothing >> Lingerie, Sl...  BRAECTFYWXNWKP7H   
13600                             ["Jewellery >> Rings"]  RNGE9HDXZYFWSFXU   

       retail_price  discounted_price  \
4787          899.0             769.0   
10573        1499.0             597.0   
6155          399.0             249.0   
2661          949.0             429.0   
13600       38106.0           27460.0   

                                                   image  \
4787   ["http://img6a.flixcart.com/image/pot-pan/c/e/...   
10573  ["http://img5a.flixcart.com/image/dress/h/p/d/...   
6155   ["http://img6a.flixcart.com/image/necklace-cha...   
2661   ["http://img6a.flixcart.com/image/bra/u/f/n/he...   
13600  ["http://img6a.flixcart.com/image/ring/k/k/m/1...   

       is_FK_Advantage_product  \
4787                     False   
10573                    False   
6155                     False   
2661                     False   
13600                    False   

                                             description       product_rating  \
4787   Buy Dailyware Tasara Kadhai 2.5 L for Rs.769 o...  No rating available   
10573  Specifications of VASTRAM Casual Sleeveless So...  No rating available   
6155   Joyeria Milan Resin Necklace - Buy Joyeria Mil...  No rating available   
2661   S4S Women's Full Coverage Bra - Buy Pink, Gree...  No rating available   
13600  BlueStone The Suneha Yellow Gold Diamond, Ruby...  No rating available   

            overall_rating          brand  \
4787   No rating available      Dailyware   
10573  No rating available        VASTRAM   
6155   No rating available  Joyeria Milan   
2661   No rating available           Pink   
13600  No rating available      BlueStone   

                                  product_specifications  
4787   {"product_specification"=>[{"key"=>"Non-stick"...  
10573  {"product_specification"=>[{"key"=>"Sleeve", "...  
6155   {"product_specification"=>[{"key"=>"Collection...  
2661   {"product_specification"=>[{"key"=>"Brand Colo...  
13600  {"product_specification"=>[{"key"=>"Pearl Type...

In [5]:
print(f"amazon shape: {amazon.shape} | flipkart shape: {flipkart.shape}")

amazon shape: (20000, 15) | flipkart shape: (20000, 15)


In [6]:
print(f"---- amazon null values ----\n"
      f"{amazon.isna().sum()}\n",
      
      f"\n---- flipkart null values ----\n"
      f"{flipkart.isna().sum()}"
      )

---- amazon null values ----
uniq_id                       0
crawl_timestamp               0
product_url                   0
product_name                  0
product_category_tree         0
pid                           0
retail_price                  0
discounted_price              0
image                         3
is_FK_Advantage_product       0
description                   2
product_rating                0
overall_rating                0
brand                      5864
product_specifications       14
dtype: int64
 
---- flipkart null values ----
uniq_id                       0
crawl_timestamp               0
product_url                   0
product_name                  0
product_category_tree         0
pid                           0
retail_price                 78
discounted_price             78
image                         3
is_FK_Advantage_product       0
description                   2
product_rating                0
overall_rating                0
brand                      58

## Preprocessing Datasets:

- Here I'm removing some columns that are not going to help in retrieving information.
- Rather than cleaning tons of rows; only going to select columns which are important, as mentioned in the pdf.

- Further I'm converting uppercase/capitalized words/strings into lowercase format.

In [7]:
def delete_redundant_columns(column_names: list, data: pd.DataFrame) -> None:
    
    """ removes redundant columns and returns the dataset"""
    
    if len(column_names) == 0:
        print(f"enter column names in the list to remove from the dataset.")
        return
    data.drop(columns=column_names, inplace=True)
    
    return

In [8]:
columns = ["uniq_id", "crawl_timestamp", "product_url", "product_category_tree","pid", "image", 
           "is_FK_Advantage_product", "description","product_rating", "overall_rating", "brand", 
           "product_specifications"
          ]

delete_redundant_columns(columns, amazon)
delete_redundant_columns(columns, flipkart)

In [9]:
amazon.head()

product_name  retail_price  discounted_price
0    Alisha Solid Women's Cycling Shorts           982               438
1    FabHomeDecor Fabric Double Sofa Bed         32143             29121
2                             AW Bellies           991               551
3    Alisha Solid Women's Cycling Shorts           694               325
4  Sicons All Purpose Arnica Dog Shampoo           208               258

In [10]:
flipkart.head()

product_name  retail_price  discounted_price
0    Alisha Solid Women's Cycling Shorts         999.0             379.0
1    FabHomeDecor Fabric Double Sofa Bed       32157.0           22646.0
2                             AW Bellies         999.0             499.0
3    Alisha Solid Women's Cycling Shorts         699.0             267.0
4  Sicons All Purpose Arnica Dog Shampoo         220.0             210.0

In [11]:
def get_lowercase_text(column: str, data: pd.DataFrame) -> None:
    
    """ converts capital/uppercase strings or text into lowercase. """
    
    if column is None:
        print(f"empty column name")
        return 
    if data.shape[0] == 0:
        print(f"empty dataset: {len(data)}")
    
    data[column] = data[column].apply(lambda x: " ".join(x.lower() for x in x.split()))
    
    return

In [12]:
get_lowercase_text('product_name', flipkart)
get_lowercase_text('product_name', amazon)

In [13]:
amazon.head()

product_name  retail_price  discounted_price
0    alisha solid women's cycling shorts           982               438
1    fabhomedecor fabric double sofa bed         32143             29121
2                             aw bellies           991               551
3    alisha solid women's cycling shorts           694               325
4  sicons all purpose arnica dog shampoo           208               258

In [14]:
flipkart.head()

product_name  retail_price  discounted_price
0    alisha solid women's cycling shorts         999.0             379.0
1    fabhomedecor fabric double sofa bed       32157.0           22646.0
2                             aw bellies         999.0             499.0
3    alisha solid women's cycling shorts         699.0             267.0
4  sicons all purpose arnica dog shampoo         220.0             210.0

In [15]:
# flipkart dataset where we can see null example in total --
# There are total 78 missing/null values in each columns retail_price and discounted_price.

# flipkart[flipkart.isna().any(axis=1)]

In [16]:
amazon.rename(columns={"product_name": "amz_product", 
                       "retail_price": "amz_retail_price", 
                       "discounted_price": "amz_discounted_price"
                      }, 
              inplace = True
             )

## Hypothesis:

- As mentioned in the problem statement that these two datasets has their products in common.
- What I want to do here is to make sure that; there are any similar products in given datasets.



- So my hypothesis here is that (also mentioned in the problem statement) there are similar products in given two datasets.
- If not we reject our hypothesis.


## Proof:

- To make sure that the given datasets has similar products, I'll use __Set__ data datastructure to check my hypothesis, I will save both dataset's product names saparately and use __intersection__ method on those two sets to get all the common products names.
- Reason behind using set is because rather than iterating through all the products and comparing one by one can lead to great computation and time complexity, instead set methods are efficient and great alternative in case of our problem.


- Since there's a lot of repetition of same product, we are going to get only one occurance of that product from both the datasets, as set intersection method gives us only the common elements from two collections.

In [17]:
amazon_prod_set = set(amazon.amz_product.to_list())
flipkart_prod_set = set(flipkart.product_name.to_list())

# -- Gives all the common products from both sets -- 
# list(amazon_prod_set.intersection(flipkart_prod_set))   

len(amazon_prod_set.intersection(flipkart_prod_set))

12604

### Note:

- Here we can see that there are total 12604 unique products that are common in both datasets, and these products are repeating themselves again and again yet with different retatil and discounted prices.
- We've proved aur hypothesis.


- Below I'm quickly going to check some products which can be available in above two sets.

In [18]:
def check_product(product_name: str, products_set: set) -> bool:
    
    """ returns a boolean value if product is present or not """
    
    if product_name == "":
        print("enter valid product name.")
        return
    if len(products_set) == 0:
        print("empty set given.")
        return
    
    if product_name in products_set:
        return True
    
    return False

In [19]:
product_names = ["alisha solid women's cycling shorts", 
                 "alisha solid women's cycling ", 
                 "faith leenise 1000137 women heels"
                ]

for prod in product_names:
    print(f"{prod} : amazon products set : {check_product(prod, amazon_prod_set)}\n"
          f"{prod} : flipkart products set : {check_product(prod, flipkart_prod_set)}\n"
         )

alisha solid women's cycling shorts : amazon products set : True
alisha solid women's cycling shorts : flipkart products set : True

alisha solid women's cycling  : amazon products set : False
alisha solid women's cycling  : flipkart products set : False

faith leenise 1000137 women heels : amazon products set : True
faith leenise 1000137 women heels : flipkart products set : True



## Combining amazon & flipkart dataset for comparison:

In [20]:
combined_data = amazon.copy()

combined_data['flipk_product'] = flipkart['product_name']
combined_data['flipk_retail_price'] = flipkart['retail_price']
combined_data['flip_discounted_price'] = flipkart['discounted_price']

# added this one extra column to see the difference between the retail prices of two datasets.

combined_data['retail_diff'] = np.where(combined_data['amz_retail_price'] == flipkart['retail_price'], 
                                        0,
                                        np.abs(combined_data['amz_retail_price'] - flipkart['retail_price'])
                                       )

combined_data.head()

amz_product  amz_retail_price  \
0    alisha solid women's cycling shorts               982   
1    fabhomedecor fabric double sofa bed             32143   
2                             aw bellies               991   
3    alisha solid women's cycling shorts               694   
4  sicons all purpose arnica dog shampoo               208   

   amz_discounted_price                          flipk_product  \
0                   438    alisha solid women's cycling shorts   
1                 29121    fabhomedecor fabric double sofa bed   
2                   551                             aw bellies   
3                   325    alisha solid women's cycling shorts   
4                   258  sicons all purpose arnica dog shampoo   

   flipk_retail_price  flip_discounted_price  retail_diff  
0               999.0                  379.0         17.0  
1             32157.0                22646.0         14.0  
2               999.0                  499.0          8.0  
3               699.0                  267.0          5.0  
4               220.0                  210.0         12.0

In [21]:
# Null data that is came along with filpkart data.

# combined_data[combined_data.isna().any(axis=1)]

## Results:

In [22]:
combined_data[combined_data['amz_product'].str.contains("muchmore alloy copper charm bracelet")]

amz_product  amz_retail_price  \
16762  muchmore alloy copper charm bracelet                -1   
16763  muchmore alloy copper charm bracelet               592   
16764  muchmore alloy copper charm bracelet               686   
16935  muchmore alloy copper charm bracelet               796   

       amz_discounted_price                         flipk_product  \
16762                     0  muchmore alloy copper charm bracelet   
16763                   349  muchmore alloy copper charm bracelet   
16764                   379  muchmore alloy copper charm bracelet   
16935                   425  muchmore alloy copper charm bracelet   

       flipk_retail_price  flip_discounted_price  retail_diff  
16762                 NaN                    NaN          NaN  
16763               600.0                  300.0          8.0  
16764               699.0                  300.0         13.0  
16935               798.0                  330.0          2.0

In [23]:
combined_data[combined_data['flipk_product'].str.contains("muchmore alloy copper charm bracelet")]

amz_product  amz_retail_price  \
16762  muchmore alloy copper charm bracelet                -1   
16763  muchmore alloy copper charm bracelet               592   
16764  muchmore alloy copper charm bracelet               686   
16935  muchmore alloy copper charm bracelet               796   

       amz_discounted_price                         flipk_product  \
16762                     0  muchmore alloy copper charm bracelet   
16763                   349  muchmore alloy copper charm bracelet   
16764                   379  muchmore alloy copper charm bracelet   
16935                   425  muchmore alloy copper charm bracelet   

       flipk_retail_price  flip_discounted_price  retail_diff  
16762                 NaN                    NaN          NaN  
16763               600.0                  300.0          8.0  
16764               699.0                  300.0         13.0  
16935               798.0                  330.0          2.0

In [24]:
combined_data[combined_data['amz_product'].str.contains("fdt women's leggings")]

amz_product  amz_retail_price  amz_discounted_price  \
28  fdt women's leggings pants               698                   362   

           flipk_product  flipk_retail_price  flip_discounted_price  \
28  fdt women's leggings               699.0                  309.0   

    retail_diff  
28          1.0

## Notes:

- As we can see that there is a repetition of a product multiple times but the retail price and discounted price is seen different at each occurances.
- There are __null values__ as well with these products because we have seen total __78 null__ values in flipkart (retail price and discounted price columns), which we have combined here to compare the products and there prices.

- I cannot impute the missing the values in repective fields, since there were no instructions according to this subject.


- Two things I consider, When we have missing data or values in dataset.  
    1. Data Engineering team should work on that if available.
    2. As a data scientist I will talk to the seniors, or data eng. team, understand reasons behind the missing data or the missing data shows a pattern that we could take into account and then I'd proceed further with my work.



__So here I'm leaving missing data as it is__

In [25]:
combined_data.shape

(20000, 7)

In [26]:
def fetch_product_info(product_name: str) -> pd.DataFrame:
    global combined_data
    
    product_name = product_name.lower()
    
    return combined_data[combined_data['amz_product'].str.contains(product_name)]

In [27]:
fetch_product_info("women's")

amz_product  amz_retail_price  \
0                    alisha solid women's cycling shorts               982   
3                    alisha solid women's cycling shorts               694   
6                    alisha solid women's cycling shorts              1198   
9                    alisha solid women's cycling shorts              1197   
11                                carrel printed women's              2281   
...                                                  ...               ...   
19786                  my addiction women's a-line dress              1293   
19787                       indibox women's a-line dress              1483   
19788                         jiiah women's sheath dress              1043   
19812  lord's red women's slingback peeptoes women heels              1280   
19881  lord's antique gold women's peeptoe heels wome...              1634   

       amz_discounted_price  \
0                       438   
3                       325   
6                       602   
9                       542   
11                     1080   
...                     ...   
19786                   934   
19787                  1136   
19788                   904   
19812                   784   
19881                  2001   

                                           flipk_product  flipk_retail_price  \
0                    alisha solid women's cycling shorts               999.0   
3                    alisha solid women's cycling shorts               699.0   
6                    alisha solid women's cycling shorts              1199.0   
9                    alisha solid women's cycling shorts              1199.0   
11                                carrel printed women's              2299.0   
...                                                  ...                 ...   
19786                  my addiction women's a-line dress              1299.0   
19787                       indibox women's a-line dress              1499.0   
19788                         jiiah women's sheath dress              1049.0   
19812  lord's red women's slingback peeptoes women heels              1299.0   
19881  lord's antique gold women's peeptoe heels wome...              1650.0   

       flip_discounted_price  retail_diff  
0                      379.0         17.0  
3                      267.0          5.0  
6                      479.0          1.0  
9                      479.0          2.0  
11                     910.0         18.0  
...                      ...          ...  
19786                  779.0          6.0  
19787                  999.0         16.0  
19788                  734.0          6.0  
19812                  650.0         19.0  
19881                 1650.0         16.0  

[3789 rows x 7 columns]